pkg/translator/prometheusremotewrite/helper.go

```diff
@@ -0,0 +1,448 @@
+
+	for i := 0; i < pt.Exemplars().Len(); i++ {
+		exemplar := pt.Exemplars().At(i)
+
+		promExemplar := &prompb.Exemplar{
+			Value:     exemplar.DoubleVal(),
+			Timestamp: timestamp.FromTime(exemplar.Timestamp().AsTime()),
+		}
+
+		exemplar.FilteredAttributes().Range(func(key string, value pdata.AttributeValue) bool {
+			promLabel := prompb.Label{
+				Name:  key,
+				Value: value.AsString(),
+			}
+
+			promExemplar.Labels = append(promExemplar.Labels, promLabel)
+
+			return true
+		})
+
+		promExemplars = append(promExemplars, *promExemplar)
+	}
+
+	return promExemplars
+}
+
+// addSingleSummaryDataPoint converts pt to len(QuantileValues) + 2 samples.
+func addSingleSummaryDataPoint(pt pdata.SummaryDataPoint, resource pdata.Resource, metric pdata.Metric, namespace string,
+	tsMap map[string]*prompb.TimeSeries, externalLabels map[string]string) {
+	time := convertTimeStamp(pt.Timestamp())
+	// sum and count of the summary should append suffix to baseName
+	baseName := getPromMetricName(metric, namespace)
+	// treat sum as a sample in an individual TimeSeries
+	sum := &prompb.Sample{
+		Value:     pt.Sum(),
+		Timestamp: time,
+	}
+	if pt.Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
+		sum.Value = math.Float64frombits(value.StaleNaN)
+	}
+	sumlabels := createAttributes(resource, pt.Attributes(), externalLabels, nameStr, baseName+sumStr)
+	addSample(tsMap, sum, sumlabels, metric)
+
+	// treat count as a sample in an individual TimeSeries
+	count := &prompb.Sample{
+		Value:     float64(pt.Count()),
+		Timestamp: time,
+	}
+	if pt.Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
+		count.Value = math.Float64frombits(value.StaleNaN)
+	}
+	countlabels := createAttributes(resource, pt.Attributes(), externalLabels, nameStr, baseName+countStr)
+	addSample(tsMap, count, countlabels, metric)
+
+	// process each percentile/quantile
+	for i := 0; i < pt.QuantileValues().Len(); i++ {
+		qt := pt.QuantileValues().At(i)
+		quantile := &prompb.Sample{
+			Value:     qt.Value(),
+			Timestamp: time,
+		}
+		if pt.Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
+			quantile.Value = math.Float64frombits(value.StaleNaN)
+		}
+		percentileStr := strconv.FormatFloat(qt.Quantile(), 'f', -1, 64)
+		qtlabels := createAttributes(resource, pt.Attributes(), externalLabels, nameStr, baseName, quantileStr, percentileStr)
+		addSample(tsMap, quantile, qtlabels, metric)
+	}
+}
+
+// copied from prometheus-go-metric-exporter
+// sanitize replaces non-alphanumeric characters with underscores in s.
+func sanitize(s string) string {
+	if len(s) == 0 {
+		return s
+	}
+
+	// Note: No length limit for label keys because Prometheus doesn't
+	// define a length limit, thus we should NOT be truncating label keys.
+	// See https://github.com/orijtech/prometheus-go-metrics-exporter/issues/4.
+	s = strings.Map(sanitizeRune, s)
+	if unicode.IsDigit(rune(s[0])) {
+		s = keyStr + "_" + s
+	}
+	if s[0] == '_' {
+		s = keyStr + s
+	}
```

Comment:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7112 seeks to eliminate this sanitization.  The current proposal is to add a feature gate to control whether to use it as I don't think anyone currently working on the project has a solid understanding of why it was added and I'm reluctant to change behavior that can impact metrics and dashboards without knowing whether it'll reintroduce long-solved and forgotten problems.

I think the feature gate could still be added here, but if this package is used outside of a collector context it may not be operable.

Reply:
Hi, once that feature gate is added, we can expose the function as `MetricsToPRW(sanitizeLabels bool, ....)`. I am happy to rebase if #7112 is merged first.

Reply:
It was my hope that we could avoid adding that option to any of the exported API surface so that it could be removed entirely if we do decide that it is unnecessary.

Reply:
I think we shouldn't be giving any stability guarantees on the module just yet, and I think it should be okay to change the interface as needed.